### DISCUSSION TITLE
5.4 release planning

### DISCUSSION THREAD

**Discussion by binste** [vega/altair#3478] [General]
We have some type hint and doc improvements as well as the removal of numpy and pandas as reqiuired dependencies (#3452). I think we could do a 5.4 release with this so that the Polars devs (cc @MarcoGorelli) could start working on the plotting API if they want to. 

I can take care of upgrading Vega-Lite to the latest version (#3479) and cutting the release. Does anyone else have a PR open they first want to get in or any objections/inputs?

**Update 07.08.2024**: See https://github.com/vega/altair/discussions/3478#discussioncomment-10235498 for the current status

**Comment by dangotbanned** [comment]
@binste 

#3455 and #3466 are both ready

#3427 I would love to get in as well, so any input on the discussions between @mattijn and myself would be very welcomed.

I do have another typing-related PR ready ~~to submit as well~~ #3480, but had been holding off until `api.py` is less busy.
*Edit 1:* Turns out that didn't have any conflicts

*Edit 2:*
## All now merged
- #3455
- #3466
- #3427
- #3480

**Comment by jonmmease** [comment]
I think https://github.com/vega/altair/pull/3479 still needs to update the vl-convert version bounds: https://github.com/vega/altair/blob/main/NOTES_FOR_MAINTAINERS.md#updating-vl-convert-version-bound. Unfortunately, Vega-Lite 5.19.0 was released a couple of days ago and I haven't done a vl-convert release with this version yet.

I probably won't have time to push through a vl-convert updated until the weekend. So there are three options:
 1. Wait until next week to do the Vega-Altair release
 2. Release against Vega-Lite 5.18.1 (which is included in vl-convert 1.5). 
 3. Someone else can try their hand at making the update in https://github.com/vega/vl-convert.

**Reply by binste** [reply]
No stress at all from my side! I just thought I'd kick off the discussion and I forgot that vl-convert needs a release first :)

Of course, if someone wants to try the update themselves that's great! Let's see, there are also some other PRs it seems which we could get into this release. And as you wrote, we can always release against VL 5.18.1 which is totally fine.

**Reply by jonmmease** [reply]
Heh, actually I was confused by my own release notes. Vega-Lite 5.19.0 is already released in vl-convert 1.5, so nothing is blocking there after all!

**Comment by dangotbanned** [comment]
I haven't started a PR for it yet, but maybe the policy for this?
- https://github.com/vega/altair/issues/3454

Thinking ahead a bit, but introducing a policy in `5.4.0` that defines some public/private boundaries could open up what changes are possible for `5.5.0`.

I think its worth considering since `5.5.0` could be 1-6 months away (guessing based on https://github.com/vega/altair/tags)

**Reply by dangotbanned** [reply]
- https://github.com/vega/altair/pull/3488

Anyone feel free to contribute, I'm going to loop back to this after https://github.com/vega/altair/discussions/3478#discussioncomment-10104219

A key part that is missing is regarding *what* is the **public API**. 
As that is [**item 1**](https://semver.org/#spec-item-1) for [SemVer](https://semver.org/) I think it will need at least a loose definition.

**Reply by dangotbanned** [reply]
@mattijn @binste @joelostblom @jonmmease

I need some help on defining the [Public API](https://semver.org/#spec-item-1) as it is the only item left for #3488 - but really needs consensus from maintainers.
Related sections for [polars](https://docs.pola.rs/development/versioning/#what-qualifies-as-a-breaking-change) and [pandas](https://pandas.pydata.org/docs/development/policies.html#python-support)

I've also been collecting prior discussions in the [**Motivation**](https://github.com/vega/altair/issues/3454#issue-2384573336) section of the original issue - which may be helpful.

**Reply by mattijn** [reply]
Maybe I'm misunderstanding, but for me the definition of the public API is quite simple. It is what is defined at the [API reference page](https://altair-viz.github.io/user_guide/api.html). So in line with what polars is stating:

> _A feature is part of the public API if it is documented in the API reference._

And I also think the order at that page is an order that defines the order from which I think is sensible.

1. [Top-Level Objects](https://altair-viz.github.io/user_guide/api.html#top-level-objects)
2. [Encoding Channels](https://altair-viz.github.io/user_guide/api.html#encoding-channels)
3. [API Functions](https://altair-viz.github.io/user_guide/api.html#api-functions)
4. [Low-Level Schema Wrappers](https://altair-viz.github.io/user_guide/api.html#low-level-schema-wrappers)

Where I can imagine that with the release of a major version we may decide to nest the low-level schema wrappers one level deeper, eg within `altair.core`.

**Reply by dangotbanned** [reply]
Thanks @mattijn, simple is good!

I'm happy with whatever you all agree on, just want to make sure the agreement is there 😄 
To give some additional context, I thought there might be more discussion around this:

@mattijn 
> Maybe I'm misunderstanding, but for me the definition of the public API is quite simple. It is what is defined at the [API reference page](https://altair-viz.github.io/user_guide/api.html).

vs 

@binste's comment here https://github.com/vega/altair/pull/3482#issuecomment-2241577342
> I'd prefer if we look at defining the public API in one go holistically by going through all elements in a systematic way.
> By doing it in one go, we can hopefully reduce frictions for end users and provide clear indication of the changes, migration guides, and a guarantee of what is public going forward.
> It will involve tradeoffs between making Altair better for future users while not breaking too many things for existing codebases and users.
> Discussing this topic with an overview of all the parts of the package hopefully allows us to get this right.

---

If I've understood both comments correctly
- one approach is the public API is already defined
- the other is that we *could* narrow that (to some degree) for the initial definition

The benefit of the latter, for example, could mean [removing specific top-level symbols](https://github.com/vega/altair/pull/3482#issuecomment-2241551416) without requiring a major version bump.

**Reply by mattijn** [reply]
If you ask me, I would say: yes

**Reply by dangotbanned** [reply]
> If you ask me, I would say: yes

I'm sorry @mattijn I don't know what you mean, could you clarify for me please?

**Reply by mattijn** [reply]
What I meant was, _yes_, the public API is already defined and _yes_, we can remove specific top-level symbols without requiring a major version bump. Eg the examples you mention in https://github.com/vega/altair/pull/3482#issuecomment-2241551416:

> The following are not accessible at the top-level:
> 
> - Mixin classes
> 	- `channels.DatumChannelMixin`
> 	- `channels.FieldChannelMixin`
> 	- `channels.ValueChannelMixin`
> - Module whose members are already accessible at top-level
> 	- `schema.mixins`
> - Internal type alias
> 	- `api.DataType`
> - Internal decorator
> 	- `schemapi.with_property_setters`
> - Internal constant (used in [private function](https://github.com/dangotbanned/altair/blob/1e0c12d2e10ddf23636233d1df6f432cccbe861d/altair/vegalite/v5/api.py#L3741))
> 	- `api.TOPLEVEL_ONLY_KEYS`

Are not mentioned at https://altair-viz.github.io/user_guide/api.html and therefore not considered public.

**Reply by mattijn** [reply]
Maybe I'm jumping too quick to conclusions as `alt.Undefined` is not mentioned at https://altair-viz.github.io/user_guide/api.html, but it actually should.

**Reply by mattijn** [reply]
Probably we should add a section on types on the API reference page, since there are a few more that can be considered useful, eg. `alt.DataFrameLike`

**Reply by dangotbanned** [reply]
> Maybe I'm jumping too quick to conclusions as `alt.Undefined` is not mentioned at [altair-viz.github.io/user_guide/api.html](https://altair-viz.github.io/user_guide/api.html), but it actually should.

https://github.com/vega/altair/blob/main/tools/generate_api_docs.py

There are others like `alt.Parameter`, really anything that doesn't subclass `SchemaBase`.

**Reply by binste** [reply]
I'm fine with using the API references in the docs to define what the public API is. But honestly, if a package exposes something in it's top-level namespace without a `_` preifx, as a user, I don't check the API references to see if it's documented. I just use it and expect that a `altair<6` in my requirements keeps me safe ;) I could imagine I'm not the only one.

As a compromise, we could flesh out what else should be added (`Parameter`, `Undefined`, `UndefinedType`, `DataFrameLike`, some more...) but refrain from removing anything from the top-level namespace, even if it was not declared in the API references, until Altair 6? What I was actually referring to in my comment which @dangotbanned quoted above, was that if during this process, we see that we want to remove some things, we could clean-up and define the public API in one go and release Altair 6 later this year.

**Reply by dangotbanned** [reply]
> What I was actually referring to in my comment which @dangotbanned quoted above, was that if during this process, we see that we want to remove some things, we could clean-up and define the public API in one go and release Altair 6 later this year.

Apologies for misunderstanding your position @binste

> I'm fine with using the API references in the docs to define what the public API is. But honestly, if a package exposes something in it's top-level namespace without a `_` preifx, as a user, I don't check the API references to see if it's documented. I just use it and expect that a `altair<6` in my requirements keeps me safe ;) I could imagine I'm not the only one.

Totally agree and this is exactly the scenario that adopting a policy is intended to avoid

> As a compromise, we could flesh out what else should be added (`Parameter`, `Undefined`, `UndefinedType`, `DataFrameLike`, some more...) but refrain from removing anything from the top-level namespace, even if it was not declared in the API references, until Altair 6? 

Building on this idea, if [those changes](https://github.com/vega/altair/discussions/3478#discussioncomment-10190114) were made *prior* to `v6`:
- Changes could be eased in with a [\_\_getattr\_\_]( https://github.com/vega/altair/issues/3454#issuecomment-2223382656) and then actually removed from the top-level in `v6`.
- Achieving the benefits of removal (not appearing in autocomplete) while not breaking existing code
	- satisfying the **MINOR** version definition in [SemVer](https://semver.org/#summary).

To approach this with extra caution, we could do something like:
- `v5.4.0` - Versioning policy introduced. We use the `v5.4.0` [API reference](https://altair-viz.github.io/user_guide/api.html) as the initial Public API.
- `v5.5.0` - First opportunity to start deprecating top-level. If we move straight to `v6.0.0`, policy is clear that it breaking and no need for deprecation.
- `v6.0.0` - Scheduled removal of any deprecations.

I think this would limit the number of *potential* surprises

**Reply by binste** [reply]
No reason to apologize! I didn't know myself what I wanted to say... ;)

**Comment by dangotbanned** [comment]
@binste I'd like to revisit a draft PR you have

- #2927 

Since https://github.com/vega/altair/pull/3431 was merged, its become clearer to me that `pylance` is including `if TYPE CHECKING: ...` imports as well. 

So this *not only* increases the number of top level symbols, but since they aren't actually present at runtime - `pylance` users would run into this issue:

![image](https://github.com/user-attachments/assets/85e32b2c-86b3-4b63-a9be-4289204c098f)

![image](https://github.com/user-attachments/assets/b0ac7049-9b02-4a74-8113-8c4f862519e3)

---

Personally, I'm happy to make this a priority if you see this as the regression that I do.

**Reply by binste** [reply]
That's quite a surprising behaviour. Do you think it's something worth reporting upstream for pylance? Feels like a bug to me. Thanks for noticing! I'd also consider this a regression we want to fix first before releasing.

Specifically for Altair, if you have the cycles to revisit autogenerating `__all__`, if this solves the problem then I think that would be great! I can't support a lot but I can probably carve out the time to review it if that works for you.

**Reply by dangotbanned** [reply]
@binste I'm glad you agree, I think I have most of it ready but will move it out of draft when I'm sure
- https://github.com/vega/altair/pull/3482

**Comment by joelostblom** [comment]
I would be interested in having https://github.com/vega/vega-lite/pull/9374 merged and there being a new Vega-Lite release that we could release Altair 5.4 against. This would make it easier to make progress on https://github.com/vega/altair/pull/3394.

This is by no means blocking from my side, but checking in with @domoritz @kanitw if either of you have time to make a VL release with https://github.com/vega/vega-lite/pull/9374 soonish or if you prefer to wait until there are more merged changes?

**Comment by dangotbanned** [comment]
Thanks everyone for all the recent reviews. 

I'll be trying to get a follow-up docs PR for `when-then-otherwise` ready soon, as mentioned in https://github.com/vega/altair/pull/3485#issue-2418667202 and https://github.com/vega/altair/pull/3427#issuecomment-2237588984.

## PRs
- #3490 
- #3492 

## Issues
- #3500

**Reply by dangotbanned** [reply]
> *  https://github.com/vega/altair/issues/3500 
> 
> can happen with this release but can happen later.

@binste 

Would anyone see a benefit in somebody who isn't me writing this into the user guide?


Not asking of anyone in particular, and I feel like I could get a PR done next week.

However it would be a good test of how useful the docstrings are.
Maybe highlighting things that are missing/too verbose after some "real-world" usage.

Things that make sense or seem helpful to me may not be universal after all

**Comment by binste** [comment]
Summary of where I think we are at but please correct me if I'm missing anything or anyone is waiting for input:

* Must: Get #3515 in so that Polars can use Altair 5.4 as a minimum dependency in https://github.com/pola-rs/polars/pull/17995
* Optional: We wait a bit to see if support for VL 5.20.1 lands in vl-convert https://github.com/vega/vl-convert/issues/176 so that we can support `react`. No rush at all, we can also release with VL 5.19 as-is. -> **Opened** https://github.com/vega/altair/pull/3525

I believe all other open points in this discussion have been addressed. #3500 can happen with this release but can happen later.

**Reply by dangotbanned** [reply]
- https://github.com/vega/altair/pull/3501

Has been ready for review but @joelostblom is still away I believe 

I'm happy to wait for the review, but this would be easy to include if anyone else can take a look? 

Closes 2 issues

**Reply by dangotbanned** [reply]
To keep things organized, I'll be keeping this meta-issue updated with any `alt.themes`-related PRs.

- https://github.com/vega/altair/issues/3519

**Comment by binste** [comment]
Thanks everyone! 5.4.0 is out now https://github.com/vega/altair/releases/tag/v5.4.0 🥳  Shoutout to @dangotbanned who shaped this release with very significant contributions and to @MarcoGorelli for getting rid of the pandas and numpy dependencies!

**Reply by dangotbanned** [reply]
Thanks @binste!

I know there is probably a lot of work to get this done, but is there any plan to continue the release notes style from [v5.3.0](https://github.com/vega/altair/releases/tag/v5.3.0)?

It seemed to be [well received](https://github.com/vega/altair/discussions/3333#discussioncomment-8972818) and we might already have screenshots/videos to reuse from the PRs

**Reply by binste** [reply]
Thanks for bringing it up. I really like the screenshots and videos myself but as you said, it can be quite some additional effort. I'd also prefer to do more frequent releases to get new features and bug fixes out sooner so I wouldn't want to make this a requirement for the person doing the release. To quote [this comment](https://github.com/vega/altair/discussions/3333#discussioncomment-9015117) by Joel:

> whoever has time to create some screenshots/animations for each release should feel free to add them to the notes (but not a requirement for the person making the release, although it's good if it happens at the same time I guess in terms of people getting notifications).

I don't plan to further edit the release notes in the next few days. If anyone wants to improve them, that would of course be amazing :)

**Reply by joelostblom** [reply]
Wow, I count 79 merged PRs! That's the most since 5.0 I believe. Thanks for making the release @binste and everyone for their contributions!

**Reply by domoritz** [reply]
Congratulations for the release. Exciting improvements and great to have a leaner Altair.
